### DISCUSSION TITLE
RetroAchievements support! Feedback wanted!

### DISCUSSION THREAD

**Discussion by rafaelvcaetano** [rafaelvcaetano/melonDS-android#1117] [General]
Hello guys!

One of the features that was thrown around a lot by the users was support for [RetroAchievements](https://retroachievements.org/). I'm happy to say that work on this has started and I am ready to show what is working so far! As of now, you can grab the latest [nightly release](https://github.com/rafaelvcaetano/melonDS-android/actions/workflows/main.yaml) and try it for yourself. You will need to have a RetroAchievements account and login within the app. To view achievement progress, click on the gear icon in the ROM entry, and then select the RetroAchievements tab as shown in the image below. If you are not logged in, you will see a button that will prompt you for your credentials so that you can login.

![image](https://user-images.githubusercontent.com/9086379/221883850-25636839-4036-48af-8ba7-bc11d771a4a9.png)

![achievement_unlock](https://user-images.githubusercontent.com/9086379/221900029-575b3e6d-51c2-4620-9a3d-67e863a631d4.gif)

**Already working**:
* [x] Login/logout with RetroAchievements
* [x] View achievements for games
* [x] Unlock achievements while playing
* [x] Softcore and Hardcore support
* [x] Rich presence
* [x] View achievement list while in-game
* [x] Notifications for primed achievements
* [x] Notifications for progress updates
* [x] Leaderboards
* [x] Multi-set support
* [x] Achievement progress in achievement list

**To be implemented**:
* No more planned features for now

If you have any feedback regarding what is already implemented, or suggestions for features that have not been listed above, leave it down in this discussion.

**Comment by drisc** [comment]
Just had a go with Pokemon Ranger which is a very touch heavy game. I prefer to play games like that using the S-Pen on my phone but the way RetroArch maps touch input it makes it unusable. MelonDS on the other hand works exactly like an actual DS screen and tracks perfectly with the point of the S-Pen. It was so much easier to play precisely and unlock an achievement using it. Can't wait for the full release with hardcore support!

**Comment by Skyrimer3d** [comment]
How do I download the nightly apk?

**Reply by drisc** [reply]
It's the artifact under the messages in a build entry from the list that is linked in the OP.

![Screenshot_20230301_101110_Firefox Nightly.jpg](https://user-images.githubusercontent.com/2449110/222109567-f014e50e-42e4-4978-8e83-aff303f1fb8e.jpg)

It's at the bottom and should be around 8mb.

**Reply by Skyrimer3d** [reply]
Found it thanks!

**Reply by HarisOfNazareth** [reply]
@drisc Could you please send the link of the nightly release. I am not a native english speaker so i am having trouble understanding what you wrote.

**Reply by HarisOfNazareth** [reply]
Rejoice all for i hath downloaded ze nightly release of ze melon-ds and it only took me abput an hour to do so. Hip hip hurray Hip hip hurray.

**Comment by Skyrimer3d** [comment]
Ok on I quick test Retro achievements seem to be working fine, I checked Newer SMB, added my RA info and it showed my currently obtained achievements with retroarch. I'll play a bit and see if they get unlocked properly, and big thanks for adding this feature!

**Comment by Skyrimer3d** [comment]
Sorry i'm going crazy about this, is there any way to disable auto-rotate or at least force portrait mode to be always active?

**Reply by drisc** [reply]
You can lock the orientation using a custom layout, the option is under Settings > Input > Layouts. Within the layout editor there is a Menu called Properties, in there is the setting for orientation.

**Comment by Skyrimer3d** [comment]
Found it thanks! Now i'll try to provide some feedback instead of annoying you with questions XP

**Comment by EgoBizarro** [comment]
Wow, great stuff, this is already working pretty well! 
Only, it doesn't seem to work when roms are 7z compressed even though melonDS supports the format. Is that not possible for some reason or will that be added later?

**Reply by Skyrimer3d** [reply]
OK I've tried this and I'm not sure it's registering correctly. I can login on my retroarch account and see the game trophies, even in Newer SMB I can see the Trophies I've already achieved using retroarch. However when I boot any of the supported games, they don't show any info on Retro achievements, and when I go to the retro achievements website it doesn't show up in my list of recently played games, something that happens inmediately with any supported game in retroarch. Can anyone suggest a DS game with very easy achievements to see if it unlocks?

**Reply by rafaelvcaetano** [reply]
Session information is not yet being set. I'm currently working on it, together with support for rich presence. There's no feedback when entering a game. You just have to trust that it's working for now :P

**Reply by Skyrimer3d** [reply]
Ah OK, so I'll try to play some games and hopefully achievements will unlock normally, glad to hear, I'll post whatever I find. One really offtopic question, is there any way to increase screen size just horizontally? And big thanks for your amazing work with this emulator!

**Reply by rafaelvcaetano** [reply]
Not yet, even though that has already been requested a lot. I still haven't found a good way to implement the UI for that.

**Reply by Skyrimer3d** [reply]
At least now I know I'm not missing the option, thanks for the reply!

**Comment by Maximdraco** [comment]
I just started a game, the Retroachievements integration seems to be working fine, my only problem is than it doesn't let me change from softcore to hardcore mode (a game mode than restric some emulator features like loading save states while playing), but apart of that it seems to be working, I haven't checked the rich presence or leaderboard yet.

**Comment by RenanSD007** [comment]
Thank you so much for working on achievement support ❤️ maybe it can get other devs interested in adding achievement support in standalone emulators :) I just don't test it on my cell phone because I've tested the emulator before and it doesn't run very well (Moto G5)

**Reply by IQuore** [reply]
"get other devs interested in adding achievement support in standalone emulators" 
I still hope someday there will be a Gameboy/GBA emulator with this achievements feature.

**Comment by Skyrimer3d** [comment]
Ok tested Sonic Rush and managed to unlock my first achievement with Melon DS 
![image](https://user-images.githubusercontent.com/99598769/222450296-78eb7323-39fd-4ef0-9369-d06c2a4ad8bc.png)
  The achievement icon appeared on the emulator when I got it, and it shows up as unlocked on the list of achievements of the game in MelonDS menu. Again, nothing shows up in my profile play list about recent games played ore recent achievements unlocked, but searching for the game on RA website showed it as unlocked, so indeed it's working as expected, hurray!

**Reply by Skyrimer3d** [reply]
OK so there's something wrong here, I've been playing more Sonic Rush and everytime I get 100 coins the achievement icon of the "this used to be harder" achievement shows up, even though I unlocked it some time ago. Also I've tried Phoenix Wright using my retroarch save, and I started the chapter from the beginning and it showed that I had unlocked an achievement, but I think I had it already since on Retroarch I was well past that point in the chapter, so it looks like it's not recognizing the achievements already unlocked and shows them again and again if you do the unlocking task once more.

**Reply by rafaelvcaetano** [reply]
That's exactly the kind of feedback I'm looking for! Thanks for reporting this! It should be fixed in the new nightly release.

**BE AWARE**: You will have to delete your app's data for the new version to work. This will be an annoying problem with future nightly releases until I come up with a good enough solution, but I hope you understand.

**Reply by oddgoo** [reply]
Build #89 Fixed the achievement popping up again for me :)

**Comment by Skyrimer3d** [comment]
No problem, I'm here to help.

**Comment by Skyrimer3d** [comment]
I've tried the "Do not load already unlocked achievements #89" build, but when I go to the retroachievements part of the game, it always says "There was a problem traying to load the achievements for this game..." in every game on the list. I've tried uninstalling and reinstalling the app completely, still the same issue.

**Comment by Skyrimer3d** [comment]
Hi, so any news about the issue I posted?

**Reply by rafaelvcaetano** [reply]
I didn't have much time, but I couldn't reproduce the problem after reinstalling the app. I'll try to investigate further. Meanwhile, try it again to make sure that you did every step correctly.

Also, could you keep everything in a single discussion by replying to this thread only? Otherwise, this conversation will be split and harder to track.

**Comment by cyruzwings** [comment]
retro achievement run well, i try few games though. 
any chance add hardcore mode on/off in achievement?
I use "Save RetroAchievements state in savestates #95"

**Reply by rafaelvcaetano** [reply]
For now, only softcore is available. Hardcore support is in development.

**Comment by EgoBizarro** [comment]
RetroAchievements don't seem to work when roms are 7z compressed even though melonDS supports the format. Is that not possible for some technical reason or will that be added later on?

**Reply by rafaelvcaetano** [reply]
It's a bug. Basically, the game hash is not being calculated properly when decompressing the ROM in memory. And I have no idea how to fix it 😔

**Reply by EgoBizarro** [reply]
Is it the same when it's in normal zip files?
Oh well, I just good myself a new SD card, may as well uncompress my DS roms I guess :)

**Reply by rafaelvcaetano** [reply]
Nope. Zip files work fine.

**Reply by EgoBizarro** [reply]
Alright, thank you!

**Reply by Maximdraco** [reply]
@rafaelvcaetano you could try to talk with some of the creators of the RA integration like Jamiras to find a way to solve the problem

**Reply by alejvilla89** [reply]
FYI this issue is also present when loading roms via retroarch which is how most consoles (with retroachievements) are played.

**Comment by IQuore** [comment]
This update is gorgeous, it's good to know that I wont be forced to use RetroArch to get NDS achievements. I really hope you'll be putting "Hardcore achievements" soon. Then I'll start playing for good, because DS on retroarch is awful.

**Reply by IQuore** [reply]
I tested NFS MW, the file is in .zip format and not only it ran well but the RetroAchievments worked well too. It has shown on my profile the game, and the details. One thing I would ask you to put is, the moment that the game is open, there could be a pop-up to show that the RA profile is connected and how much achievements I have, like "(1/21 achievements)".

**Reply by alejvilla89** [reply]
I’m with you there, waiting patiently for hardcore achievement support! And very grateful that rafaelvcaetano has taken the time to work on this.

**Comment by Seydie** [comment]
I am so happy to read this! I specifically bought a tablet + S pen for DS achievements but couldn't get it to work properly on RetroArch. Works great already, can't wait for the hardcore support to fully play Pokémon Ranger games.

**Comment by Captainquill** [comment]
This is just amazing. Having known that the retroarch cores for ds on Android are pretty average melon ds standalone getting retroachievements support is great news. Tried it, works as expected. Just a few things that I wish to be implemented in the future is hardcore achievement support because I pretty much always play hardcore and secondly a resolution upscale feature just like drastic Will make me ditch all other options for this amazing emulator. Best of luck.

**Comment by nighto** [comment]
Sad iPhone user here. Any chance this is backported to main melonDS? Thanks! :)

**Reply by rafaelvcaetano** [reply]
It should be relatively easy. The worst part is designing (and implementing) a UI that makes sense for this feature in a desktop environment.

**Comment by Captainquill** [comment]
Any update regarding the hardcore achievements implementation..? Thank you.

**Reply by rafaelvcaetano** [reply]
It's still ongoing! I took some time to make some changes in order to improve the code so that this whole feature works without issues. Unfortunately, those changes took longer than expected, but it's almost done.

**Reply by Captainquill** [reply]
You are doing gods work 😊... If I may can I ask when exactly is the update going to be made available to download. I understand things take time but I am emphasizing this much because I work onboard a ship and within next few days I will be joining back my work so I wont be able to download it if it gets released later. I just want to download it and take it with me with my ds games so that I can have a great time onboard. Thanks in advance.

**Reply by alejvilla89** [reply]
You need to maintain an internet connection to earn retroachievements. If you can earn achievements while onboard the ship then shouldn’t you also be able to download the update?

**Reply by drisc** [reply]
When you are on a ship you can still get internet but achievement unlocks are a couple of kilobytes compared to the megabytes you would need to use to download an emulator update. Shipboard wifi can also be very expensive if you go over your data cap.

**Reply by Captainquill** [reply]
Exactly just as drisc said, I get 200 mb of data for one month that too is painfully slow.. So you could imagine downloading a 10 mb apk is kind of a big deal... On the other hand even if I play the game for achievements online I have noticed that the data usage is not more than a few kilobytes.

**Reply by rafaelvcaetano** [reply]
Actually, provided you have already downloaded the achievement list for the games you want (either by launching the ROM or accessing the achievement list), you can unlock achievements while offline and once you connect to the internet, it will communicate with the RetroAchievements' server to unlock them online. But it won't track play time. Though, I'm not sure if this is a good practice and I'll have to check with them to see if this is fine before the next proper release.

**Reply by alejvilla89** [reply]
Thanks all for the explanation. I was unaware. Though I feel bad I caused the conversation to go off-track. 

rafaelvcaetano, Captainquill had asked if you had settled on a release date or at least a window for the hardcore achievements update since he is due to board a ship in the coming days and would love having it before boarding. Any insight would be appreciated!

**Comment by kensher0** [comment]
thank you so much for this, you are a hero.
and I can't wait for hardcore mode support...

**Comment by Captainquill** [comment]
Good day.. Was pleasantly surprised to find out that the latest nighty build just released has a button for activation hardcore achievements in settings.. But don't know why after updating the retroachievements has stopped loading completely. Now when we click on the cog and the retroachievements tab it says.. There was a problem trying to load achievemts. Is anybody else having the same issue. Thanks

**Reply by rafaelvcaetano** [reply]
You're very fast! Ahahaha. I was about to post an update with more information, but had to leave for a bit. To fix this, you need to clear app data or reinstall the app. For the nightly builds this will be necessary because otherwise I would spend a lot of time writing migrations.

**Reply by Captainquill** [reply]
Hahaha... Thanks to me coming to this page and checking for an update every hour 😂... Finally that I got it very happy now.. Thanks a lot.

**Reply by Captainquill** [reply]
Hardcore achievements are unlocking properly... I have one more question regarding the update. Is there a way to check the achievements list from inside the game now..

**Reply by rafaelvcaetano** [reply]
Not yet. But it's planned.

**Comment by oddgoo** [comment]
Weeping with Joy

![image](https://user-images.githubusercontent.com/4585118/235800607-567d9202-9bb5-45e0-8d97-7d399f281b0d.png)

For the latest nightly, Everything seems to be working great in the first hour of testing (Samsung S23Ultra)

**Reply by alejvilla89** [reply]
I can also confirm that the hardcore achievements seem to be working as they should. Thank you so much!!! The Android options for earning DS retroachievements have been pretty bad (for lack of better words) and you just did a ton for this segment of the community. Great work! I really appreciate it.

**Comment by Captainquill** [comment]
Hello Rafael... Can you implement an unlock sound just like retroarch or aethersx2 in this? I am sure others will also appreciate the satisfying sound of achievement unlocking. Thanks.

**Reply by EgoBizarro** [reply]
If so, with an option to disable it, though, please :)

**Comment by Captainquill** [comment]
Hello Rafael... Hope everything's well on your side. Just commenting this to make sure. I see there's has not been a single update or a comment from tour side for past almost 2 months. Thanks.

**Reply by rafaelvcaetano** [reply]
Hello! Thanks for checking out! The truth is that I took some time to play Tears of the Kingdom 😅 Since I don't have a lot of time, I couldn't keep doing both things at the same time, but I still tried to respond to the most urgent comments. However, I am almost done with it, normal development will resume shortly 🙂

**Reply by EgoBizarro** [reply]
Hehe, that's not a game to be rushed. Take your time and have fun :)

**Reply by Captainquill** [reply]
😅 Good to know you are also a fellow Zelda fan... Take your time with that amazing game. We will be patiently waiting the next update. Thank you.

**Comment by Captainquill** [comment]
Hello Rafael.. I don't know if you are aware that the DSi games also have support for retroachievements now since last month using the bizhawk emulator on windows... Would like to know as melonds already supports playing DSi games on android can the achievements support be brought for DSi games also on this amazing emulator. Thanks in advance.

**Comment by oddgoo** [comment]
Enjoy TOTK and thank you for your hard work! :)

On Thu, 29 Jun 2023 at 23:17, Captainquill ***@***.***> wrote:

> Hello Rafael.. I don't know if you are aware that the DSi games also have
> support for retroachievements now since last month using the bizhawk
> emulator on windows... Would like to know as melonds already supports
> playing DSi games on android can the achievements support be brought for
> DSi games also on this amazing emulator. Thanks in advance.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rafaelvcaetano/melonDS-android/discussions/1117#discussioncomment-6315337>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABC7NHXDBT7QIG3AK63IC2LXNV55VANCNFSM6AAAAAAVKZZAV4>
> .
> You are receiving this because you commented.Message ID:
> <rafaelvcaetano/melonDS-android/repo-discussions/1117/comments/6315337@
> github.com>
>


-- 
*Cuauhtemoc Moreno*
*Creative Coder and Game Designer*

**Comment by DragonCurse4** [comment]
Will achievements be added to the Windows version?

**Reply by RenanSD007** [reply]
melonDS-emu/melonDS#1146

**Reply by nighto** [reply]
Unfortunately upstream developer doesn't see value in RetroAchievements (commented "what's the point" etc.) so it looks like they won't be adding this. Eventually someone could fork it out and make it happen, or just use another emulator on Windows such as BizHawk https://github.com/TASEmulators/BizHawk

**Comment by nighto** [comment]
Question from an iPhone user who's eager to give it a try but doesn't have an Android: does melonDS works fine on a gaming Android based device such as GPD XD? I wouldn't buy an Android phone but it could be an excuse to get an Android based gaming device :)

**Reply by rafaelvcaetano** [reply]
I've just looked at the specs of the GPD XD, and while the emulator will run on it, I don't think it will provide a smooth experience.

**Reply by nighto** [reply]
Thanks @rafaelvcaetano. Do you suggest a specific Android based device focused on gaming?

**Comment by JobiMorkos** [comment]
I want a feature where you could see the achievement on the corner where for example, When the player is on the boss the "No Hit the Boss" achievement icon shows up on the corner indicating there some achievement during that boss fight but if the player gets hit at the boss fight, The achievement icon on the corner disappear. 

(Also this emulator runs pretty well on my device.)

**Comment by Tardigrade-nx** [comment]
Hey,

Great work on the RetroAchievements support!
MelonDS is now the only alternative for fullspeed DS emulation with RA on T618 devices (Retroid Pocket 3+, Flip, ...).
They work fine for me in hardcore more. However, they don't seem to unlock in softcore mode. They load fine, just don't unlock. Is it just me? The 1st post says Hardcore and Softcore support, so I was wondering. ;-)
I'm using the latest nightly, Build #116.

**Reply by rafaelvcaetano** [reply]
I have been trying to replicate your issue, but I've had no success. Are you having issues on a specific game? Or is it a general issue?

**Reply by Tardigrade-nx** [reply]
I've had the issue with Super Princess Peach, but if it's just me, I'll do more testing.

Update: tried again with latest nightly (beta-1.9.2), softcore RA unlock fine. 😊

**Comment by RenanSD007** [comment]
Any updates? It would be nice to add support for Leaderboards :)

**Comment by GheorgheMihai** [comment]
Hello. Yesterday I tried to play on MelonDS with Retroachievements, but it said there was no internet conection, even if I was on WiFi.I then tried to log out and log in, but I got the message 'There was a problem when trying to login'. Is there a solution to this problem?

**Reply by rafaelvcaetano** [reply]
It looks like the guys at RetroAchievements made some changes which broke the integration. For now, there is no solution. Keep a look on #1261 for updates regarding this issue.

**Reply by GheorgheMihai** [reply]
Thanks for the answer. I thought something was wrong on my part.

**Comment by kensher0** [comment]
Hi @rafaelvcaetano .
 Is there any future plans to add achievements to DSi games?

**Reply by rafaelvcaetano** [reply]
This is now available on the nightly releases.

**Comment by CringyBoi42069** [comment]
Some of the games that are listed with achievements on RetroAchievements don't show achievements on melonDS.

**Comment by Kirais** [comment]
RetroAchievement is working like a charm in melon ds. Thanks a lot for adding this feature. Is there any plan for adding leaderboard support? That would be even more awesome.

**Reply by rafaelvcaetano** [reply]
This is now available on the nightly releases.

**Comment by jmaydev-personal** [comment]
Any possibility of updating the Retro Achievements functionality to support the new multi-set feature being rolled out currently? Thanks!

**Reply by rafaelvcaetano** [reply]
Now available on the nightly release

**Reply by jmaydev-personal** [reply]
Working perfectly for me on the latest update, appreciate your work, thanks!

**Comment by Jonathan-Pacheco** [comment]
This emulator is awesome, congratulations! RetroAchievements works fine, but some achievements are missing a tracker to show how much you still need to get them. For example: “Treasure chests in some area – 2/5”. Recently, I loaded my save in another emulator just to see this information hehehehehe.

**Reply by rafaelvcaetano** [reply]
This is now available on the nightly releases.

**Reply by DFelten** [reply]
Thanks, the indicator is working great! 

Could you please also add the progress to the achievement list? 

Currently it seems that only the progress indicator appears in the top left corner, but the progress is not visible within the achievement list. So it’s not possible to check which achievement I’m currently working on and the progress.

**Reply by rafaelvcaetano** [reply]
Now available on the latest nightly

**Reply by DFelten** [reply]
Thanks, it works great and looks good!

**Comment by Zayr64** [comment]
Any news on the possibility of adding Multiset for Retro Achievements?

**Reply by rafaelvcaetano** [reply]
Multi-set support is now available in the nightly release

**Reply by Brucasco** [reply]
Thanks a lot!. I just found out about the nightly apk yesterday so i will switch from the Play Store one to this one!.

**Comment by Brucasco** [comment]
I just wanted to thank you for implementing the RA's multi set feature also on MelonDS!

**Comment by Greatscoot** [comment]
Hey, I'm having an issue with earning an achievement in Castlevania: Dawn of Sorrow. Specifically, beating bosses no damage without MP use. 

When I finish the fight, and get their soul, the achievement is supposed to pop up. But it doesn't, the challenge indicator just disappears, no achievement earned. Here's a video demonstrating. https://youtu.be/JbWbCGtB_z0?si=ZXWJqn2Rp9ZGvxrq

Talked to achievement dev, he claims this is an issue with the emulator, not the achievement.

**Reply by rafaelvcaetano** [reply]
Do you have a save file that allows me to easily access the boss fight?

**Reply by Greatscoot** [reply]
I do!

File 2 is at the savepoint before the boss. https://drive.google.com/file/d/1lppLor6RtVzZEtUxshh8GwGoK7xL7dEa/view?usp=drive_link

**Reply by Greatscoot** [reply]
Here's a savestate right before you're supposed to get the achievement, that way you don't have to do the boss damageless. :p It should proc when you get the soul.

https://drive.google.com/file/d/1W5HQ5c06tT4idKP7B6EX-M6TGKvEhlIg/view?usp=sharing

**Reply by rafaelvcaetano** [reply]
Thanks for providing al this information. I've finally found the issue and pushed a fix to the nightly release.

**Comment by Mahonzu** [comment]
Thank you for listening to all feedback!
I have an issue where achievements for Pokemon: Platinum aren't loading. It says to check internet connection, but all other games work fine. Same ROM dump gives achievements fine in retroarch. 
I tried Nightly version, which fixes the issue, however the first achievement (for getting the first badge) doesn't unlock. 
Save file: https://drive.google.com/file/d/11w4lHng7bJyGVvw8y9XL4_38Cwk5hS7u/view?usp=sharing
Just enter the gym and spam Bubble attack, should oneshot everyone

**Reply by rafaelvcaetano** [reply]
Are you using hardcore mode?

**Reply by Mahonzu** [reply]
I do

**Reply by Mahonzu** [reply]
Feel free to ask for more info/details, I'll be happy to help as much as I can

**Comment by edsonbezerradasilva901-ops** [comment]
I wanted to ask you when a version of Melonds will be released that includes an achievement unlock sound after unlocking a RetroAchievements achievement, because all previous versions of Melonds don't have this sound. So I'd like to know when this update will be released.
